New component observer/cfgardenobserver (#33727)
open-telemetry/opentelemetry-collector-contrib: extension/observer/cfgardenobserver/README.md

| Status        |           |
| ------------- |-----------|
| Stability     | [development]  |
| Distributions | [] |
| Issues        | [![Open issues](https://img.shields.io/github/issues-search/open-telemetry/opentelemetry-collector-contrib?query=is%3Aissue%20is%3Aopen%20label%3Aextension%2Fcfgardenobserver%20&label=open&color=orange&logo=opentelemetry)](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues?q=is%3Aopen+is%3Aissue+label%3Aextension%2Fcfgardenobserver) [![Closed issues](https://img.shields.io/github/issues-search/open-telemetry/opentelemetry-collector-contrib?query=is%3Aissue%20is%3Aclosed%20label%3Aextension%2Fcfgardenobserver%20&label=closed&color=blue&logo=opentelemetry)](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues?q=is%3Aclosed+is%3Aissue+label%3Aextension%2Fcfgardenobserver) |
| [Code Owners](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#becoming-a-code-owner)    | [@crobert-1](https://www.github.com/crobert-1) |

[development]: https://github.com/open-telemetry/opentelemetry-collector#development
<!-- end autogenerated section -->

> Cloudfoundry is a PaaS which has implemented their own container technology. The component in charge of providing a local API in each node (also known as "diego-cell") is named Garden.

Reviewer: ```suggestion
Cloudfoundry is a platform as a service (PaaS) which has implemented their own container technology. The component in charge of providing a local API in each node (also known as "diego-cell") is named Garden.
```

The `cfgarden_observer` looks at the current host to discover garden containers.

> ### Configuration

Reviewer: It would be good to include a full configuration example here. We should also make it clear which variables are required and which are optional.

Config options are usually in a list, rather than sections. It may be good to convert these into a list to be more uniform with other READMEs.

> # CF Garden Observer Extension

Reviewer: Nit: We may want to expand Cloud Foundry here instead of just using the initials, to make it more clear what we're referring to.

| [Code Owners](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#becoming-a-code-owner)    | [@crobert-1](https://www.github.com/crobert-1), [@cemdk](https://www.github.com/cemdk), [@tomasmota](https://www.github.com/tomasmota), [@m1rp](https://www.github.com/m1rp), [@jriguera](https://www.github.com/jriguera) |

[development]: https://github.com/open-telemetry/opentelemetry-collector#development
<!-- end autogenerated section -->

> Cloudfoundry is a platform as a service (PaaS) which has implemented their own container technology. The component in charge of providing a local API in each node (also known as "diego-cell") is named Garden.

Reviewer: ```suggestion
Cloud Foundry (CF) is a platform as a service (PaaS) which has implemented their own container technology. The component in charge of providing a local API in each node (also known as "diego-cell") is named Garden.
```

> The `cfgarden_observer` looks at the current host to discover garden containers.

Reviewer: From what I can tell, `Garden` should be capitalized in all documentation and comments. (Code references can remain lower case.)

## Example Config

```yaml
extensions:
  cfgarden_observer:

>     # url of the garden socket, default to unix:///var/vcap/data/garden/garden.sock 

Reviewer: ```suggestion
    # url of the garden socket, defaults to unix:///var/vcap/data/garden/garden.sock 
```

    # determines how often to look for changes in endpoints.
    refresh_interval: 30s

receivers:
  receiver_creator:
    watch_observers: [cfgarden_observer]
    receivers:
      prometheus_simple:
        rule: type == "container" && name == "myapp"
        config:
          metrics_path: /metrics
          endpoint: '`endpoint`'
```

### Configuration

| Name             | Type   | Default                                  | Docs                                                   |
|------------------|--------|------------------------------------------|--------------------------------------------------------|
| refresh_interval | string | 60s                                      | Determines how often to look for changes in endpoints. |
| endpoint         | string | unix:///var/vcap/data/garden/garden.sock | The endpoint to connect to the garden API.             |

### Endpoint Variables

Endpoint variables exposed by this observer are as follows.

| Variable  | Description                                                                                |
|-----------|--------------------------------------------------------------------------------------------|

> | type      | `"container"`                                                                              |

Reviewer: The description here isn't immediately clear, can you clarify? Is this simply saying the value for the  exposed `type` key will always be `container`?

Reviewer 2: Yes, it will always be `container`. We just did the same as the host observer here, do you want me to change both to be clearer or none of them? https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/extension/observer/hostobserver/README.md?plain=1#L35

Reviewer: Thanks for the reference. My personal preference is to lean on the side of extra information, so I like having the extra description here.
